lib/Arch/X86/Arch.cpp

```diff
@@ -788,6 +788,55 @@ static void DecodeOperand(Instruction &inst, const xed_decoded_inst_t *xedd,
   }
 }
 
+static uint64_t BytesToCallBits(const uint8_t *bytes) {
+  // The pattern we're looking for is 5 bytes long.
+  uint64_t bits = 0;
+  bits = (bits << 8) | static_cast<uint32_t>(bytes[0]);
+  bits = (bits << 8) | static_cast<uint32_t>(bytes[1]);
+  bits = (bits << 8) | static_cast<uint32_t>(bytes[2]);
+  bits = (bits << 8) | static_cast<uint32_t>(bytes[3]);
+  bits = (bits << 8) | static_cast<uint32_t>(bytes[4]);
+  return bits;
+}
+
+static bool TryDecodeIdioms(Instruction &inst) {
+  // Check for CALL+POP idiom used to retrieve PC.
+  if (inst.bytes.size() == 6) {
+    const auto bytes = reinterpret_cast<const uint8_t *>(inst.bytes.data());
+    const auto call_bits = BytesToCallBits(bytes);
+
+    // Check that we have a CALL pointing one instruction ahead of where we are
+    // now.
+    if (call_bits != 0xe800000000) {
+      return false;
+    }
+
+    // Check whether we have a POP instruction to retrieve the return address
+    // that we just pushed to the stack.
+    Instruction pop_inst;
+    if (!inst.arch->DecodeInstruction(inst.pc + 5, &inst.bytes[5], pop_inst) ||
```

Comment:
Not sure whether this is appropriate since we're within the `DecodeInstruction` path already. I could avoid this (with a bit more mess) by calling `DecodeXED` and using the XED structures directly.

Reply:
I think there's an issue here where it's converting a `const char *` to a `std::string_view`, which in turn relies on `strlen`. You want to give it a one-byte view into the `inst.bytes` instead.

Reply:
I would also recomment being more selective about the `pop_inst.function`, as theoretically this could match on other `POP` variants.

Reply:
> I think there's an issue here where it's converting a `const char *` to a `std::string_view`, which in turn relies on `strlen`. You want to give it a one-byte view into the `inst.bytes` instead.

Why is that an issue? I'm just taking an index into `inst.bytes` which is a `std::string` so that should be a valid C-string.

Reply:
> I would also recomment being more selective about the `pop_inst.function`, as theoretically this could match on other `POP` variants.

What `POP` functions would it not apply for? I had a look at the stuff in `POP.cpp` and from what I can see, they would all have the "get pc" effect when combined with the call.

Reply:
@pgoodman